pkg/controllers/node/termination/terminator/terminator.go

```diff
@@ -98,28 +114,26 @@ func (t *Terminator) Drain(ctx context.Context, node *v1.Node) error {
 }
 
 func (t *Terminator) Evict(pods []*v1.Pod) {
```

Comment:
What about just updating our Evict() function to pull in all pods waiting termination, grouping them, and then checking if the group is empty. If the group isn't empty, then we'd just pass the pods that haven't started termination yet to the eviction queue. 

I think realistically, this allows us to drop the check on `podsWaitingEvictionCount > 0` and should reduce the amount of change required to get this functionality

Reply:
@jonathan-innis thanks for taking a look at this. Happy to work on some improvements.

I could use some clarification on a few points.

>What about just updating our Evict() function to pull in all pods waiting termination, grouping them, and then checking if the group is empty. If the group isn't empty, then we'd just pass the pods that haven't started termination yet to the eviction queue.

That would break with the current behavior where `Drain(...)` selects the pods that need to be evicted, and `Evict(...)` groups them and queues them up. (Not necessarily a bad thing, but I was trying to color within the existing lines when I wrote this code.)

In order to pull/query for not-yet-evicted pods in `Evict(...)`, we would need to call `nodeutil.GetPods(ctx, t.kubeClient, node)`. That would require a function signature change (`Evict(ctx context.Context, node *v1.Node)`) and would create a slight misalignment with how the k8s API works. We'd be calling for an eviction on a node, instead of pods.. so to speak. Is that something we'd be ok with?

>I think realistically, this allows us to drop the check on podsWaitingEvictionCount > 0 and should reduce the amount of change required to get this functionality

Is that check not required? I'm not particularly familiar with the `finalize` function that calls `Drain`, but it looks like the error returned is used to halt progress on the termination: https://github.com/kubernetes-sigs/karpenter/blob/f1a2d42b5051585503035aa0883f4cb2cd8cf9d6/pkg/controllers/node/termination/controller.go#L90-L93

Reply:
> That would break with the current behavior where `Drain(...)` selects the pods that need to be evicted, and `Evict(...)` groups them and queues them up. (Not necessarily a bad thing, but I was trying to color within the existing lines when I wrote this code.)

Totally, though I think this is a reasonable change to make, given what we are trying to achieve here. If anything, I think this allows you to actually just remove the `Evict` function entirely and just collapse all of the logic into the `Drain()` function. Inside of the drain, you can just try evicting a set of pods until all of the pods are fully terminated. The way I had thought about doing it was something like this

```
func (t *Terminator) Drain(ctx context.Context, node *v1.Node) error {
	pods, err := nodeutil.GetPods(ctx, t.kubeClient, node)
	if err != nil {
		return fmt.Errorf("listing pods on node, %w", err)
	}
        podGroups := t.groupPodsByPriority(pods)
        waitingOnEviction := false
        for _, group := range podGroups {
           if len(group) > 0 {
               t.evictionQueue.Add(lo.Filter(pods, func(p *v1.Pod, _ int) bool { return podutil.IsEvictable(p) }))
               waitingOnEviction = true
           }
       }
      if waitingOnEviction {
        return NewNodeDrainError(fmt.Errorf("%d pods are waiting to be evicted", len(pods)))
     }
}
        

func (t *Terminator) groupPodsByPriority(pods []*v1.Pod) [][]*v1.Pod {
  // 1. Prioritize noncritical pods, non-daemon pods https://kubernetes.io/docs/concepts/architecture/nodes/#graceful-node-shutdown
	var criticalNonDaemon, criticalDaemon, nonCriticalNonDaemon, nonCriticalDaemon []*v1.Pod
	for _, pod := range pods {
		if pod.Spec.PriorityClassName == "system-cluster-critical" || pod.Spec.PriorityClassName == "system-node-critical" {
			if podutil.IsOwnedByDaemonSet(pod) {
				criticalDaemon = append(criticalDaemon, pod)
			} else {
				criticalNonDaemon = append(criticalNonDaemon, pod)
			}
		} else {
			if podutil.IsOwnedByDaemonSet(pod) {
				nonCriticalDaemon = append(nonCriticalDaemon, pod)
			} else {
				nonCriticalNonDaemon = append(nonCriticalNonDaemon, pod)
			}
		}
	}
	return [][]*v1.Pod{criticalNonDaemon, criticalDaemon, nonCriticalNonDaemon, nonCriticalDaemon}
}
```